design/ratelimit-design-new.md

```diff
@@ -0,0 +1,475 @@
+    burst: 20
+  # global defines global rate limiting properties for the virtual host or route.
+  global:
+    # rateLimitDescriptors defines the descriptors to be generated and sent to the
+    # external Rate Limiting Service (RLS) for a rate limit decision.
+    rateLimitDescriptors:
+    # This descriptor is generated only if the x-steve-ratelimit header
+    # is present on the request.
+    - items:
+        # adds a descriptor entry with a key of "generic_key" and a value
+        # of "s1".
+        - genericKey:
+            value: s1
+
+        # adds a descriptor entry with a key of "remote_address" and a value
+        # equal to the trusted address from x-forwarded-for.
+        - remoteAddress: {}
+
+        # adds a descriptor entry with a key of "steve-ratelimit"
+        # and a value equal to the value of the "x-steve-ratelimit" header.
+        - requestHeader:
+            descriptorKey: steve-ratelimit
+            headerName: x-steve-ratelimit
+        
+    # This descriptor is always generated since it's just a static value.
+    - items:
+        - genericKey:
+            value: s1
+```
+
+This API maps closely to Envoy's rate limiting API.
+
+For local rate limiting, the rate limit itself is defined inline in the `RateLimitPolicy`, as token bucket properties.
+Since local rate limits apply *per Envoy*, each pod in the Envoy daemon set will have a token bucket with these properties.
+
+For global rate limiting, per Envoy's API, the `RateLimitPolicy` only defines a list of descriptors to send to an external RLS for a given request.
+All [descriptor entries defined by Envoy](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#config-route-v3-ratelimit-action) will be supported except `metadata` and `dynamic_metadata`.
+
+### Global RLS ExtensionService
+When using global rate limiting, first, an `ExtensionService` must be defined for the RLS with the cluster-level details for the RLS itself.
+For example:
+```yaml
+apiVersion: projectcontour.io/v1alpha1
+kind: ExtensionService
+metadata:
+  namespace: projectcontour
+  name: ratelimit
+spec:
+  protocol: h2
+  services:
+    - name: ratelimit
+      port: 8081
+  timeoutPolicy:
+    # sets  the "timeout" property on the Envoy rate limit service
+    # config.
+    response: 50ms
+```
+
+### HTTPProxy Changes
+
+The HTTPProxy struct adds new fields `spec.virtualhost.rateLimitServer`, `spec.virtualhost.rateLimitPolicy`, and `spec.routes[].rateLimitPolicy`:
+
+```yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: proxy
+  namespace: projectcontour
+spec:
+  virtualhost:
+    fqdn: local.projectcontour.io
+    tls:
+      secretName: local-tls
+    # rateLimitServer optionally defines a non-default RLS to use
+    # for this HTTPProxy. This field can only be specified for root
+    # TLS-enabled HTTPProxies.
+    rateLimitServer:
```

Comment:
Sorry if this is digging backward to the ext-authz extension but is this decoupled from the `rateLimitPolicy` for any specific reason? I noticed it is decoupled there too. What advantages are there in having the extension ref etc. a separate top level field in the vhost config from the policy? (why not have them all under a common `ratelimit` struct?)

Reply:
Is it so it's easily overridable by the global one in the contour bootstrap config?

Reply:
`rateLimitPolicy` can be defined at either the vhost level or at the route level, whereas `rateLimitServer` can only apply for a vhost. That said, we *could* have a parent `rateLimit` struct for vhost that contains both the `rateLimitServer` and `rateLimitPolicy` just for organization, if we wanted. Though, it would add even more nesting to the YAML.